rice-datanode/data_transfer_server.cc

```diff
@@ -0,0 +1,266 @@
+	uint64_t bytesInBlock = proto.minbytesrcvd();
+	//num bytes sent
+	uint64_t bytesSent = proto.maxbytesrcvd();
+	
+	if (rpcserver::write_delimited_proto(sock, response_string)) {
+		LOG(INFO) << "Successfully sent response to client";
+	} else {
+		LOG(INFO) << "Could not send response to client";
+	}
+	
+	// read packets of block from client
+	bool last_packet = false;
+	std::string block_data;
+	while (!last_packet) {
+		asio::error_code error;
+		uint32_t payload_len;
+		rpcserver::read_int32(sock, &payload_len);
+		uint16_t header_len;
+		rpcserver::read_int16(sock, &header_len);
+		PacketHeaderProto p_head;
+		rpcserver::read_proto(sock, p_head, header_len);
+		last_packet = p_head.lastpacketinblock();
+		uint64_t data_len = p_head.datalen();
+		uint32_t checksum_len = payload_len - 4 - data_len;
+		std::string checksum (checksum_len, 0);
+		std::string data (data_len, 0);
+		size_t len = sock.read_some(asio::buffer(&checksum[0], checksum_len), error);
+		if (len != checksum_len || error) {
+			LOG(ERROR) << "Failed to read checksum for packet " << p_head.seqno();
+			break;
+		}
+		len = sock.read_some(asio::buffer(&data[0], data_len), error);
+		if (len != data_len || error) {
+			LOG(ERROR) << "Failed to read packet " << p_head.seqno();
+			break;
+		}
+		if (!last_packet && data_len != 0) {
+			block_data += data;
+		}
+		// ack packet
+		PipelineAckProto ack;
+		ack.set_seqno(p_head.seqno());
+		ack.add_reply(SUCCESS);
+		std::string ack_string;
+		ack.SerializeToString(&ack_string);
+		if (rpcserver::write_delimited_proto(sock, ack_string)) {
+			LOG(INFO) << "Successfully sent ack to client";
+		} else {
+			LOG(INFO) << "Could not send ack to client";
+		}
+	}
+	
+	LOG(INFO) << "Writing data to disk: " << block_data;
+	if (!fs.allocateBlock(header.baseheader().block().blockid(), block_data)) {
+		LOG(ERROR) << "Failed to allocate block " << header.baseheader().block().blockid();
+	} else {
+		dn.blockReceived(header.baseheader().block().blockid());
+	}
+
+
+	//TODO set proto source to this DataNode, remove this DataNode from
+	//	the proto targets, and send this proto along to other
+	//	DataNodes in targets
+	//TODO read in a response (?)
+	//TODO send packets to targets
+}
+
+void TransferServer::processReadRequest(tcp::socket& sock) {
+	OpReadBlockProto proto;
+	if (rpcserver::read_delimited_proto(sock, proto)) {
+		LOG(INFO) << "Op a read block proto";
+		LOG(INFO) << proto.DebugString();
+	} else {
+		ERROR_AND_RETURN("Failed to op the read block proto.");
+	}
+	std::string response_string;
+
+	buildBlockOpResponse(response_string);
+	if (rpcserver::write_delimited_proto(sock, response_string)) {
+		LOG(INFO) << "Successfully sent response to client";
+	} else {
+		LOG(INFO) << "Could not send response to client";
+	}
+
+	uint64_t blockID = proto.header().baseheader().block().blockid();
+	bool success;
+	std::string block = this->fs.getBlock(blockID, success);
+	if (!success){
+		LOG(ERROR) << "Failure on fs.getBlock";
```

Comment:
(unless you have observed directly the memcpy, i think only the header is copied)